### PR TITLE
Restore onChange for calendar

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue #14306 .

### Summary of Changes
- Restore the onChange property
- Add an extra check that the field belongs to a form before attaching an `on submit` listener

### Testing Instructions
Apply patch
Edit protostar template and insert somewhere inside the body:
```html
<?php echo JHtml::_('calendar', JFactory::getDate()->format('Y-m-d'), 'date', 'date', '%Y-%m-%d', array('onChange' => "alert('it works')")); ?>
```
An alert with a message `it works` should appear each time you change the date in the calendar

### Documentation Changes Required
None, this is a bug

### CODE REVIEW

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on cc5773ce20e7cd712b197ba4fcb04686d8184d40<br /><br />Tested successfully and thank you :) saves my life.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14311">issues.joomla.org/tracker/joomla-cms/14311</a>.</sub>

File: media/system/js/fields/calendar.js [hunk 1]
```diff
@@ -1058,9 +1058,10 @@
 
 			if (!instance) {
 				instance = new JoomlaCalendar(element);
-				instance.inputField.form.addEventListener('submit', function () {
-					instance.setAltValue();
-				});
+				if (instance.inputField.form)
```
Comment (yvesh) [inline media/system/js/fields/calendar.js]: braces please ;)

**Comment by chmst** [comment]
Can you give me a hint how to get rid of the time?
There should be a attribute "showtime='false'" but it also does not work<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14311">issues.joomla.org/tracker/joomla-cms/14311</a>.</sub>

**Comment by dgt41** [comment]
Try this:
```php
	<?php echo JHtml::_('calendar', JFactory::getDate()->format('Y-m-d'), 'date', 'date', '%Y-%m-%d', array(
			'onChange' => "alert('it works')",
			"showTime" => false,
			"todayBtn" => true,
			"weekNumbers" => false,
			"fillTable" => true,
			"singleHeader" => false,

	)); ?>
```

File: media/system/js/fields/calendar.js (hunk 1 above)
Comment (dgt41) [inline media/system/js/fields/calendar.js]: this is not the only instance in this file, maybe a new PR for all the CS work?

**Comment by franz-wohlkoenig** [comment]
I have tested this item :red_circle: unsuccessfully on e50b176676a98de705fef3021fbf1c0d610d4b3e<br /><br />Insert `<?php echo JHtml::_('calendar', JFactory::getDate()->format('Y-m-d'), 'date', 'date', '%Y-%m-%d', array('onChange' => "alert('it works')")); ?>` in `/index.php` in template "protostar" after `<div class="body" id="top">
` (line 141) got:<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14311">issues.joomla.org/tracker/joomla-cms/14311</a>.</sub>

**Comment by dgt41** [comment]
@franz-wohlkoenig try in another position, e.g. inside the main after `<main id="content" role="main" class="<?php echo $span; ?>">`

File: libraries/cms/html/html.php [hunk 2]
```diff
@@ -1063,6 +1070,8 @@ public static function calendar($value, $name, $id, $format = '%Y-%m-%d', $attri
 			'helperPath'   => $helperPath,
 			'localesPath'  => $localesPath,
 			'direction'    => $direction,
+			'onchange'     => $onchange
```
Comment (zero-24) [inline libraries/cms/html/html.php]: Can we have a comma here and remove the clean line? ;)

**Comment by franz-wohlkoenig** [comment]
@dgt41 same:
bildschirmfoto 2017-03-03 um 08 31 29

**Comment by dgt41** [comment]
@franz-wohlkoenig what is the problem, as the field renders successfully?

**Comment by chmst** [comment]
@dgt41 
thanks agail, it works but with ' around the booleans
```
	$attribs = array(
			'onChange' => "alert('it works')",
			"showTime" => 'false',
			"todayBtn" => 'true',
			"weekNumbers" => 'false',
			"fillTable" => 'true',
			"singleHeader" => 'false',
		);
	echo JHtml::_('calendar', JFactory::getDate()->format('Y-m-d'), 'date', 'date', '%Y-%m-%d', $attribs); ?>
```

**Comment by chmst** [comment]
@dgt41 
thanks agail, it works but with ' around the booleans
```
	$attribs = array(
			'onChange' => "alert('it works')",
			"showTime" => 'false',
			"todayBtn" => 'true',
			"weekNumbers" => 'false',
			"fillTable" => 'true',
			"singleHeader" => 'false',
		);
	echo JHtml::_('calendar', JFactory::getDate()->format('Y-m-d'), 'date', 'date', '%Y-%m-%d', $attribs); ?>
```
@dgt41 
thanks again, it works but with ' around the values
```
	$attribs = array(
			'onChange' => "alert('it works')",
			"showTime" => 'false',
			"todayBtn" => 'true',
			"weekNumbers" => 'false',
			"fillTable" => 'true',
			"singleHeader" => 'false',
		);
	echo JHtml::_('calendar', JFactory::getDate()->format('Y-m-d'), 'date', 'date', '%Y-%m-%d', $attribs); ?>
```

**Comment by infograf768** [comment]
Works fine here with
```php
<?php echo JHtml::_('calendar', JFactory::getDate()->format('Y-m-d'), 'date', 'date', '%Y-%m-%d', array(
			'onChange' => "alert('it works')",
			"showTime" => false,
			"todayBtn" => true,
			"weekNumbers" => false,
			"fillTable" => true,
			"singleHeader" => false,

	)); ?>
```
screen shot 2017-03-03 at 08 55 38

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on d0879d23ab9d650508d9b0a5416533344d56671f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14311">issues.joomla.org/tracker/joomla-cms/14311</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
@dgt41:
e45e4328-ffeb-11e6-8b01-2f105746a397

**Comment by dgt41** [comment]
@chmst @infograf768 so do we need the quotes or not?

**Comment by dgt41** [comment]
@franz-wohlkoenig that is the expected result, to test try to change the date and see if you get a popup with the text `it works`

**Comment by chmst** [comment]
@dgt41 , @infograf768 
see the screens (don't mind the css .. I'm working on it)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14311">issues.joomla.org/tracker/joomla-cms/14311</a>.</sub>

**Comment by infograf768** [comment]
@dgt41 
I did not need the quotes here.

**Comment by franz-wohlkoenig** [comment]
@dgt41 ah, pop-up: i don't get them to work. So all fine as others tested successfully.

**Comment by infograf768** [comment]
@chmst 

Which screens?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14311">issues.joomla.org/tracker/joomla-cms/14311</a>.</sub>

**Comment by chmst** [comment]
Strange, I need them: screen shot 2017-03-03 at 08 14 47screen shot 2017-03-03 at 08 14 47<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14311">issues.joomla.org/tracker/joomla-cms/14311</a>.</sub>

**Comment by wilsonge** [comment]
Merged :)

**Pull request merged**